src/NumericalAlgorithms/Spectral/SwshInterpolation.cpp

```diff
@@ -0,0 +1,557 @@
+                                      m);
+      evaluate_swsh_at_l_min_plus_one(
+          make_not_null(&current_cached_harmonic_l_plus_one),
+          current_cached_harmonic, m);
+    } else {
+      evaluate_swsh_m_recurrence_at_l_min(
+          make_not_null(&current_cached_harmonic), m);
+      evaluate_swsh_at_l_min_plus_one(
+          make_not_null(&current_cached_harmonic_l_plus_one),
+          current_cached_harmonic, m);
+    }
+    if (Spin == 0 and m == 0) {
+      cached_base_harmonic = current_cached_harmonic;
+    }
+    clenshaw_sum(interpolated, current_cached_harmonic,
+                 current_cached_harmonic_l_plus_one, goldberg_modes, m);
+  }
+  // perform the Clenshaw sums over m < 0.
+  for (int m = -1; m >= -static_cast<int>(l_max_); --m) {
+    if (m == -1 and Spin == 0) {
+      current_cached_harmonic = cached_base_harmonic;
+    }
+    if (abs(Spin) >= abs(m)) {
+      direct_evaluation_swsh_at_l_min(make_not_null(&current_cached_harmonic),
+                                      m);
+      evaluate_swsh_at_l_min_plus_one(
+          make_not_null(&current_cached_harmonic_l_plus_one),
+          current_cached_harmonic, m);
+    } else {
+      evaluate_swsh_m_recurrence_at_l_min(
+          make_not_null(&current_cached_harmonic), m);
+      evaluate_swsh_at_l_min_plus_one(
+          make_not_null(&current_cached_harmonic_l_plus_one),
+          current_cached_harmonic, m);
+    }
+    clenshaw_sum(interpolated, current_cached_harmonic,
+                 current_cached_harmonic_l_plus_one, goldberg_modes, m);
+  }
+}
+
+template <int Spin>
+void SwshInterpolator::interpolate(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, Spin>*> interpolated,
+    const SpinWeighted<ComplexDataVector, Spin>&
+        libsharp_collocation) noexcept {
+  SpinWeighted<ComplexModalVector, Spin> libsharp_modes;
+  libsharp_modes.set_data_ref(raw_libsharp_coefficient_buffer_.data(),
+                              raw_libsharp_coefficient_buffer_.size());
+  swsh_transform(l_max_, 1, make_not_null(&libsharp_modes),
+                 libsharp_collocation);
+  SpinWeighted<ComplexModalVector, Spin> goldberg_modes;
+  libsharp_to_goldberg_modes(make_not_null(&goldberg_modes), libsharp_modes,
+                             l_max_);
+  interpolate(interpolated, goldberg_modes);
+}
+
+template <int Spin>
+void SwshInterpolator::direct_evaluation_swsh_at_l_min(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, Spin>*> harmonic,
+    const int m) const noexcept {
+  const auto& clenshaw_factors = cached_clenshaw_factors<Spin>(l_max_);
+  harmonic->data() =
+      // this is the right order of the casts, other orders give the wrong
+      // answer
+      // NOLINTNEXTLINE(misc-misplaced-widening-cast)
+      clenshaw_factors.harmonic_at_l_min_prefactors[static_cast<size_t>(
+          m + static_cast<int>(l_max_))] *
+      (std::complex<double>(1.0, 0.0) * gsl::at(cos_m_phi_, abs(m)) +
```

Comment:
Why not write this as `std::complex<double>(gsl::at(cos_m_phi_, abs(m), (m >= 0 ? 1.0 : -1.0) * gsl::at(sin_m_phi_, abs(m)))`? I think you've got similar code I've seen before where rather than multiplying and adding complexes you could just create a complex from the doubles

Reply:
The multiplying and adding here is required because I'm ending up with a `ComplexDataVector` from individual components that are `DataVector`s.